src/main/java/bio/terra/workspace/db/DataReferenceDao.java

```diff
@@ -0,0 +1,120 @@
+import bio.terra.workspace.generated.model.DataReference;
+import bio.terra.workspace.generated.model.DataRepoSnapshot;
+import bio.terra.workspace.service.datareference.create.exception.InvalidDataReferenceException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.openapitools.jackson.nullable.JsonNullable;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DataReferenceDao {
+
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+  private final ObjectMapper objectMapper;
+
+  @Autowired
+  public DataReferenceDao(
+      WorkspaceManagerJdbcConfiguration jdbcConfiguration, ObjectMapper objectMapper) {
+    this.jdbcTemplate = new NamedParameterJdbcTemplate(jdbcConfiguration.getDataSource());
+    this.objectMapper = objectMapper;
+  }
+
+  public String createDataReference(
+      UUID referenceId,
+      UUID workspaceId,
+      String name,
+      JsonNullable<UUID> resourceId,
+      JsonNullable<String> credentialId,
+      String cloningInstructions,
+      JsonNullable<DataReference.ReferenceTypeEnum> referenceType,
+      JsonNullable<DataRepoSnapshot> reference) {
+    String sql =
+        "INSERT INTO workspace_data_reference (workspace_id, reference_id, name, resource_id, credential_id, cloning_instructions, reference_type, reference) VALUES "
+            + "(:workspace_id, :reference_id, :name, :resource_id, :credential_id, :cloning_instructions, :reference_type, cast(:reference AS json))";
+
+    Map<String, Object> paramMap = new HashMap<>();
+    paramMap.put("workspace_id", workspaceId.toString());
+    paramMap.put("reference_id", referenceId.toString());
+    paramMap.put("name", name);
+    paramMap.put("cloning_instructions", cloningInstructions);
+
+    if (credentialId.isPresent()) {
+      paramMap.put("credential_id", credentialId.get());
+    } else paramMap.put("credential_id", null);
+
+    if (resourceId.isPresent()) {
+      paramMap.put("resource_id", resourceId.get().toString());
+      paramMap.put("reference_type", null);
+      paramMap.put("reference", null);
+    } else {
+      if (referenceType.isPresent() && reference.isPresent()) {
+        paramMap.put("resource_id", null);
+        paramMap.put("reference_type", referenceType.get().getValue());
+        try {
+          paramMap.put("reference", objectMapper.writeValueAsString(reference));
+        } catch (JsonProcessingException e) {
+          throw new InvalidDataReferenceException("Could not write data reference to database");
+        }
+      } else
+        throw new InvalidDataReferenceException(
+            "DataReference must contain either a resource ID or a reference type and a reference");
+    }
```

Comment:
You could use a ternary operator to simplify this a bit: something like
`paramMap.put("reference_type", resourceId.isPresent() ? null : referenceType.get());` and handling the exception from calling `get` when `referenceType` is missing